src/zcash/Address.cpp

```diff
@@ -36,6 +38,47 @@ SproutPaymentAddress SproutSpendingKey::address() const {
     return viewing_key().address();
 }
 
+//! Sapling
+SaplingFullViewingKey SaplingExpandedSpendingKey::full_viewing_key() const {
+    uint256 ak;
+    uint256 nk;
+    librustzcash_ask_to_ak(ask.begin(), ak.begin());
+    librustzcash_nsk_to_nk(nsk.begin(), nk.begin());
+    return SaplingFullViewingKey(ak, nk, ovk);
+}
+
+SaplingExpandedSpendingKey SaplingSpendingKey::expanded_spending_key() const {
+    return SaplingExpandedSpendingKey(PRF_ask(*this), PRF_nsk(*this), PRF_ovk(*this));
+}
+
+SaplingFullViewingKey SaplingSpendingKey::full_viewing_key() const {
+    return expanded_spending_key().full_viewing_key();
+}
+
+SaplingIncomingViewingKey SaplingFullViewingKey::in_viewing_key() const {
+    uint256 ivk;
+    librustzcash_crh_ivk(ak.begin(), nk.begin(), ivk.begin());
+    return SaplingIncomingViewingKey(ivk);
+}
+
+SaplingSpendingKey SaplingSpendingKey::random() {
+    return SaplingSpendingKey(random_uint256());
+}
+
+boost::optional<SaplingPaymentAddress> SaplingIncomingViewingKey::address(diversifier_t d) const {
+    uint256 pk_d;
+    if (librustzcash_check_diversifier(d.data())) {
+        librustzcash_ivk_to_pkd(this->begin(), d.data(), pk_d.begin());
+        return SaplingPaymentAddress(d, pk_d);
+    } else {
+        return boost::none;
+    }
```

Comment:
Just as a general style issue, I prefer to minimize how much state I need to remember when reading code sequentially, all else being equal. In this case, if the ``librustzcash_check_diversifier`` test were inverted &ndash;i.e. returning ``boost::none`` in the ``if`` body rather than the ``else`` body&ndash; then you would be able to conclude when reading sequentially that the invalid-diversifier case was already dealt with, when reading the code for the valid case, rather than either having to remember that this will be handled by the ``else`` or read non-sequentially. (This conclusion also depends on the code for the invalid case being shorter and less complicated.)

Does not block.